1660 create home page programs block

candywang opened the pull request:
Created the `ProgramsBlock` component for the home page using the block `ProgramCard` component

candywang committed:
Create ProgramsBlock component directory, subdirectory, and files

candywang committed:
Export ProgramsBlock component

candywang committed:
Add images for programs

candywang committed:
Create stories Template component for ProgramsBlock component

candywang committed:
Create classes for ProgramsBlock component

candywang committed:
Create ProgramsBlock component

candywang committed:
Merge remote-tracking branch 'origin/main' into 1660-create-home-page-programs-block

candywang committed:
Integrate ProgramsBlock component onto home page

richardxia reviewed (commented):
Nice, this code looks pretty clean, so I don't think I have any comments about that!

Regarding the actual behavior of the components across different screen sizes, I have the following comments:

## Non-mobile, but smaller screen
On a smaller screen, but still "laptop-sized", the cards end up wrapping in an awkward way:

<img width="1333" alt="Screen Shot 2020-10-26 at 10 46 31 AM" src="https://user-images.githubusercontent.com/1002748/97208593-8a601880-1778-11eb-835d-705b120d9889.png">

I think this is being caused by your `max-width: 410px` on the actual card component itself. What you need to do is have the parent element here constrain the size of the card. The parent's job is to define how much space the children are allowed to have, and the children are forced to adjust to that space.

Your choice of flexbox here means that you should be setting the [`flex-basis`](https://developer.mozilla.org/en-US/docs/Web/CSS/flex-basis) property (or the third component of the [`flex`](https://developer.mozilla.org/en-US/docs/Web/CSS/flex) shorthand based on some number that is proportional to the actual width of our block here. This math actually gets tricky because flexbox doesn't have the concept of the gap (or "gutter") between flex items, so you'd have to subtrack the space for the gap before dividing by 3.

Alternatively, you could switch to using grid, where you can force three columns per row and have the columns shrink accordingly.

## Mobile view

The margins are a bit too small between the title and the first card as well as between the other cards:

<img width="492" alt="Screen Shot 2020-10-26 at 10 52 23 AM" src="https://user-images.githubusercontent.com/1002748/97209366-7a950400-1779-11eb-9abf-ead7232e2b13.png">

I know we don't have any mobile designs here, but the general rule of thumb of whitespace in designs is that there should be larger spacing separating different components than the spacing within the component. For example, there should be more space separating two cards than space separating the image from the text within a single card. In your current code, you have a 20px margin separating the image from its title within the same card, but there's 0 margin between the bottom of a description of one card and the image of the next card, which is a bit confusing because it makes it look like the title and description are tied to the card below rather than the card above.

You may want to take a look at some of the mobile designs for other components to get some ideas for how much spacing we should have.

candywang committed:
Merge remote-tracking branch 'origin/main' into 1660-create-home-page-programs-block

candywang committed:
Change programs theme to dark mode

candywang committed:
Change title font color to white and component background to black

candywang committed:
Change display to grid, add row gap and margin top in media query

candywang committed:
Fix lint issues

candywang commented:
I made the following changes per @richardxia suggestions and based on the new Figma mobile design:
- Changed program cards' theme to be in "dark mode" where the background is black and the font color is white
- Switched flexbox to grid display
- In the media query for the program card block I added `grid-row-gap: 30px;` for spacing between each program card and `margin-top: 30px;` property for spacing above the component that holds all the program cards so that there is visual space between the title and program cards block


Tablet view:
<img width="852" alt="Screen Shot 2020-11-01 at 6 45 53 PM" src="https://user-images.githubusercontent.com/39474473/97825122-ad0b9900-1c72-11eb-9f6a-687f42c0f7f0.png">

Large mobile view:
<img width="430" alt="Screen Shot 2020-11-01 at 6 46 09 PM" src="https://user-images.githubusercontent.com/39474473/97825130-b1d04d00-1c72-11eb-8b37-dc18a8b6cabe.png">

Small mobile view:
<img width="332" alt="Screen Shot 2020-11-01 at 6 46 22 PM" src="https://user-images.githubusercontent.com/39474473/97825133-b3017a00-1c72-11eb-9bd2-4f7e62034d8b.png">

davidagustin reviewed (dismissed):
No inconsistencies found, LGTM

candywang committed:
Change ProgramsBlock component theme to "dark"

richardxia commented:
On a laptop/desktop resolution screen, it looks like your cards are not aligned to the grid:

<img width="1361" alt="Screen Shot 2020-11-01 at 11 17 41 PM" src="https://user-images.githubusercontent.com/1002748/97840238-85c7c280-1c98-11eb-90e7-41fe07f72dde.png">

This is most noticeable on slightly narrower windows, where you can eventually get it to the point where all the cards are right next to each other with no margin in between. I think you're probably missing some grid column gap in that innermost grid, which would provide a fixed, minimum spacing between the elements.

richardxia commented on src/components/grid-aware/ProgramsBlock/ProgramsBlock.module.css:
```diff
@@ -0,0 +1,76 @@
+
+@media (--tablet-and-down) {
+  .bleedWrapper {
+    grid-template-columns: 40px 1fr 40px;
```
It looks like in the mobile designs for this component, it's actually a 20px margin on the sides, not 40px.

richardxia commented on src/components/grid-aware/ProgramsBlock/ProgramsBlock.module.css:
```diff
@@ -0,0 +1,76 @@
+
+@media (--tablet-and-down) {
+  .programsWrapper {
+    grid-row-gap: 30px;
```
According to the designs, it looks like this gap should actually be closer to 40px.

davidagustin replied:
The component as a whole has 30px gap

![Screen Shot 2020-11-03 at 4 06 05 PM](https://user-images.githubusercontent.com/30575095/98053981-aefa6700-1dee-11eb-865a-385f2859b290.png)

![Screen Shot 2020-11-03 at 4 06 17 PM](https://user-images.githubusercontent.com/30575095/98054007-be79b000-1dee-11eb-9f9a-99392a6093fd.png)

![Screen Shot 2020-11-03 at 4 05 58 PM](https://user-images.githubusercontent.com/30575095/98054018-c9344500-1dee-11eb-8148-c3505de946db.png)

The textbox in relation to the picture of another card is around 40px although
![Screen Shot 2020-11-03 at 4 06 54 PM](https://user-images.githubusercontent.com/30575095/98054048-d9e4bb00-1dee-11eb-9ae9-c0913d4a6016.png)

![Screen Shot 2020-11-03 at 4 06 49 PM](https://user-images.githubusercontent.com/30575095/98054056-dea96f00-1dee-11eb-8b0d-be5194b171c5.png)

candywang replied:
I noticed that the spacings between the three cards are inconsistent. The first gap between the first and second card has a spacing of 41px and the second gap between the second and third card has a spacing of 46px. I made a comment on the design to ask @derekfidler for his thoughts on this. In the meantime, I'll change this gap to 40px.

davidagustin replied:
No the spacings are good @cw323 , you're looking at the textbox and not the component as a whole

richardxia commented on src/components/grid-aware/ProgramsBlock/ProgramsBlock.module.css:
```diff
@@ -0,0 +1,76 @@
+
+.programsWrapper {
+  display: grid;
+  grid-template-columns: repeat(3, 1fr);
```
Regarding my comment from last night, if you add a `grid-column-gap` property here, it should allow all the program cards to be aligned to the parent grid.

richardxia reviewed (commented):
This is looking pretty close! I think if you address the last few things I just commented on, this will be good to merge in!

candywang committed:
Decrease side margins in .bleedWrapper media query, add grid-column-gap property in .programsWrapper

candywang commented:
I fixed the issue for the missing `grid-column-gap: 30px` property in the `.programsWrapper` selector. The cards are now aligned to the grid regardless of laptop/desktop resolution screen size. I also changed the gap on the mobile view to have a 40px spacing.

**Narrow laptop view**:
<img width="831" alt="Screen Shot 2020-11-03 at 5 08 55 PM" src="https://user-images.githubusercontent.com/39474473/98057980-7495c780-1df8-11eb-8f6c-ce9c649a289d.png">

**Large mobile view**:
<img width="288" alt="Screen Shot 2020-11-03 at 5 09 57 PM" src="https://user-images.githubusercontent.com/39474473/98057985-78294e80-1df8-11eb-848e-984cd204bfa5.png">

davidagustin reviewed (approved):
LGTM after discussion with Richard

richardxia reviewed (approved):
Cool, this LGTM! Thanks for all the hard work on this!

candywang commented:
Thank you! 🤓
